### PR TITLE
Enable cross_entropy loss to handle dense targets

### PR DESCRIPTION
**Pull request by AtomicVar**
## Proposed changes

Dense targets means probabilities or one-hot encodings.

Currently, the `cross_entropy` loss only supports targets to be class indices. However, in PyTorch, the [`CrossEntropyLoss`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) supports targets to be both class indices and probabilities (or one-hot encodings). In TensorFlow, two seperate cross-entropy loss layers are provided: [`CategoricalCrossentropy`](https://www.tensorflow.org/api_docs/python/tf/keras/losses/CategoricalCrossentropy) and [`SparseCategoricalCrossentropy`](https://www.tensorflow.org/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy), to support two different types of targets.

This PR enhances the `cross_entropy` loss to support both sparse targets (class indices) and dense targets (probabilities or one-hot encodings). It can infer what type of targets from the shape, like PyTorch does.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -32,20 +32,59 @@ def cross_entropy(
...
+        >>> logits = mx.array([[2.0, -1.0], [-1.0, 2.0]])
+        >>> targets = mx.array([0, 1])
+        >>> cross_entropy(logits, targets)
+        array([0.0485873, 0.0485873], dtype=float32)
+
+        >>> # Probabilities (or one-hot vectors) as targets
+        >>> logits = mx.array([[2.0, -1.0], [-1.0, 2.0]])
+        >>> targets = mx.array([[0.9, 0.1], [0.1, 0.9]])
+        >>> cross_entropy(logits, targets)
+        array([0.348587, 0.348587], dtype=float32)
```
Comment (awni) [inline python/mlx/nn/losses.py:63]: ❤️

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -62,7 +101,9 @@ def cross_entropy(
 
     # Apply weights if provided
     if weights is not None:
-        if weights.shape != targets.shape:
+        if (targets_as_probs and weights.shape != _drop_dim(targets.shape, axis)) or (
+            not targets_as_probs and weights.shape != targets.shape
+        ):
```
Comment (awni) [inline python/mlx/nn/losses.py:106]: I think you could simplify this and check `weights.shape == loss.shape` ?

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -62,7 +101,9 @@ def cross_entropy(
...
             raise ValueError(
                 f"Weights with shape {weights.shape} is not the same as "
                 f"targets with shape {targets.shape}."
```
Comment (awni) [inline python/mlx/nn/losses.py:109]: Please update this error message since now the weight should have the same shape as the loss but not necessarily the target.

File: python/mlx/nn/losses.py [hunk 4]
```diff
@@ -32,20 +32,59 @@ def cross_entropy(
...
+    # Check shapes in two cases: targets as class indices and targets as probabilities
+    if (targets_as_probs and targets.shape != logits.shape) or (
+        not targets_as_probs and targets.shape != _drop_dim(logits.shape, axis)
+    ):
+        raise ValueError(
+            f"Targets shape {targets.shape} does not match logits shape {logits.shape}."
+        )
+
+    if targets_as_probs:
+        score = mx.sum(logits * targets, axis=axis)  # mx.inner doesn't support axis
```
Comment (awni) [inline python/mlx/nn/losses.py:84]: nit: `mx.inner` is unlikely to support axis since numpy is the same, so the comment is not necessary

Comment (awni) [reply python/mlx/nn/losses.py:84]: You could (and probably should) use `mx.tensordot` though.

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for adding this! I left just a few comments, please address then we can merge.

File: python/mlx/nn/losses.py (hunk 4 above)
Comment (AtomicVar) [reply python/mlx/nn/losses.py:84]: Thanks for your advice, but I cannot figure out how to use `mx.tensordot` to do `mx.sum(a * b, axis=axis)`. For example, if both `a` and `b` has shape `[2, 3, 4]`, with `mx.sum(a * b, axis=1)` I can get the desired output with shape `[2, 4]` (`3` is reduced). But with `mx.tensordot(a, b, axes=[[1], [1]])`, it produces shape `[2, 4, 2, 4]`. I would like to use something like `np.einsum` to do `mx.einsum("ijk,ijk->ik", a, b)` but MLX doesn't have a Einsum.

**Comment by AtomicVar** [comment]
@awni Thank you for taking the time to review the PR. I've pushed a new commit to address to problems, except the `mx.tensordot` thing. Let me know if this is acceptable.

File: python/mlx/nn/losses.py (hunk 4 above)
Comment (awni) [reply python/mlx/nn/losses.py]: Indeed, sorry `tensordot` is not the right call, sorry for the bad suggestion. Let's keep the sum reduce as you have it actually, that is simple and likely fine perf wise.

**Review by awni — APPROVED** [review]
Thanks for the addition, looks great!!
